src/bootstrap/test.rs

```diff
@@ -352,6 +352,55 @@ impl Step for Rls {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct RustAnalyzer {
+    stage: u32,
+    host: TargetSelection,
+}
+
+impl Step for RustAnalyzer {
+    type Output = ();
+    const ONLY_HOSTS: bool = true;
+    const DEFAULT: bool = true;
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.path("src/tools/rust-analyzer")
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        run.builder.ensure(Self { stage: run.builder.top_stage, host: run.target });
+    }
+
+    /// Runs `cargo test` for rust-analyzer
+    fn run(self, builder: &Builder<'_>) {
+        let stage = self.stage;
+        let host = self.host;
+        let compiler = builder.compiler(stage, host);
+
+        builder.ensure(tool::RustAnalyzer { compiler, target: self.host }).expect("in-tree tool");
+
+        let path = "src/tools/rust-analyzer";
+        let mut cargo = tool::prepare_tool_cargo(
+            builder,
+            compiler,
+            Mode::ToolStd,
+            host,
+            "test",
+            path,
+            SourceType::InTree,
+            &["rust-analyzer/in-rust-tree".to_owned()],
+        );
+
+        let dir = builder.src.join(path);
+        cargo.env("CARGO_WORKSPACE_DIR", &dir);
```

Comment:
Why are you overriding this? I think cargo will do different feature / dependency resolution for the different workspace, no?

Reply:
this is likely also what's breaking `x.py check` in your IDE

Reply:
cargo doesn't set that, unfortunately: https://github.com/rust-lang/cargo/issues/3946

RA issue: https://github.com/rust-analyzer/expect-test/issues/33

Reply:
oh I see, this is only read by rust-analyzer, not cargo itself. can you add that as a comment?